### PR TITLE
feat: [BT-6045]: Extract script from pipeline and generalize

### PR DESCRIPTION
**Pull request by mbatchelor**
## Checklist
- [ ] I've documented the changes in the PR description.
- [ ] I've tested this change either in PR or local environment.
- [ ] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>

- Feature build: `trigger feature-build`
- 
  Specific builds
  - `trigger manager`
  - `trigger dms`
  - `trigger ng_manager`
  - `trigger cvng `
  - `trigger cmdlib`
  - `trigger template_svc`
  - `trigger events_fmwrk_monitor`
  - `trigger event_server`
  - `trigger change_data_capture`
  -  Trigger multiple builds together: For eg: `trigger dms, manager`
- Immutable delegate `trigger publish-delegate`
</details>

## [Latest PR Check Triggers](https://github.com/harness/harness-core/blob/develop/.github/pull_request_template.md)

<details>
  <summary>PR Check triggers</summary>
You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ut0, ut1`

- Compile: `trigger compile`
- CodeformatCheckstyle: `trigger checkstylecodeformat`
  - CodeFormat: `trigger codeformat`
  - Checkstyle: `trigger checkstyle`
- MessageMetadata: `trigger messagecheck`
- File-Permission-Check: `trigger checkpermission`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- Trigger CommonChecks: `trigger commonchecks`
- PMD: `trigger pmd`
- Copyright Check: `trigger copyrightcheck`
- Feature Name Check: `trigger featurenamecheck`
- UnitTests-ALL: `trigger uts`
- UnitTests-0: `trigger ut0`
- UnitTests-1: `trigger ut1`
- UnitTests-2: `trigger ut2`
- UnitTests-3: `trigger ut3`
- UnitTests-4: `trigger ut4`
- UnitTests-5: `trigger ut5`
- UnitTests-6: `trigger ut6`
- UnitTests-7: `trigger ut7`
- UnitTests-8: `trigger ut8`
- UnitTests-9: `trigger ut9`
- CodeBaseHash: `trigger codebasehash`
- CodeFormatCheckstyle: `trigger checkstylecodeformat`
- SonarScan: `trigger ss`
- GitLeaks: `trigger gitleaks`
- Trigger all Checks: `trigger smartchecks`
- Go Build: `trigger gobuild`
- Validate_Reviews: `trigger review`
- Module Dependency Check: `trigger mdc`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

### Rebase Instructions:
To rebase your branch onto the latest changes in the target branch, simply use the following command in a comment: `/rebase`

## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/55834)

### CODE REVIEW

File: scripts/jenkins/module-dependency-check.sh [hunk 1]
```diff
@@ -0,0 +1,199 @@
...
+        { "add": "'"$2"'" }
+      ] } }'
+  )
+
+  if  [[ "$response" -ne 204 ]] ; then
+    echo "Failed to add label to $1 - Response code: $response"
+    if [[ "$response" -eq 404 ]] ; then
+      echo "404 response indicates that user $JIRA_USERNAME does not have permissions to update labels on $1"
+    fi
+  else
```
Comment (ivan-levelops) [inline scripts/jenkins/module-dependency-check.sh:33]: we could have other error messages, better to specify the actual success code to report the success and instead report a generic error on any other not handled response code

File: scripts/jenkins/module-dependency-check.sh [hunk 2]
```diff
@@ -0,0 +1,199 @@
...
+
+  if [ "$4" == "null" ]; then
+    echo "Missing module"
+    exit 1
+  fi
+
+  ##############################
+  # Read PR diff into temp file
+  ##############################
+  resp=$(curl -L -s "https://patch-diff.githubusercontent.com/raw/harness/harness-core/pull/$1.diff" --write-out '%{http_code}' --output pr.diff)
```
Comment (ivan-levelops) [inline scripts/jenkins/module-dependency-check.sh:58]: this needs to support both github (GH) and harness code (HC)

File: scripts/jenkins/module-dependency-check.sh [hunk 3]
```diff
@@ -0,0 +1,199 @@
...
+        echo "Could not download ${ghprbPullId}.diff from github"
+        exit 1
+    fi
+  else
+    echo "curl command malformed."
+    exit 1
+  fi
+
+  # Extract changed files
+  cat pr.diff | grep "^diff " | sed 's/diff --git a\///' | sed 's/ b\/.*$//' > $tmpdiff
```
Comment (ivan-levelops) [inline scripts/jenkins/module-dependency-check.sh:72]: might need adjustments for HC

File: scripts/jenkins/module-dependency-check.sh [hunk 4]
```diff
@@ -0,0 +1,199 @@
...
+    else
+      if [ "$3" == "null" ]; then
+        echo "Missing label to apply"
+        exit 1
+      fi
+
+      if [ "$4" == "null" ]; then
+        echo "Missing module"
+        exit 1
+      fi
```
Comment (ivan-levelops) [inline scripts/jenkins/module-dependency-check.sh:96]: This doesn't really matter in here as `$4` is not being used (and it has already been checked for `null` at the begining.. same goes for the null check for `$3` in the previous lines).

are we missing the use fr `$4` in here?

File: scripts/jenkins/module-dependency-check.sh (hunk 2 above)
Comment (mbatchelor) [reply scripts/jenkins/module-dependency-check.sh:57]: Future...

File: scripts/jenkins/module-dependency-check.sh (hunk 3 above)
Comment (mbatchelor) [reply scripts/jenkins/module-dependency-check.sh:71]: Future...

**Pull request merged**